### ISSUE TITLE
TPS running low

### ISSUE THREAD

**Issue by ExWhyZedd0** [Tschipcraft/dynamiclights#13]
### Describe the bug

hello, its me again. Still love your work !!
so, i tried the 1.7 ver but its make my server tps running low. when i change it back to 1.6 ver, my tps back normal.
i dont know why its happening, can you try to do something about it ?
thank you :)

**Comment by Tschipcraft** [comment]
Thank you for the bug report, I will take a look at this.

**Comment by Ninty9** [comment]
Hey, I have no idea if this user was having the same issue that i had, but i noticed my tps was running really low around a specific users base, and it turned out that there where _90.000_ marker entities at their base. found out from the tags that they where from your mod.

Another user said that it Might be because they get unloaded as soon as they spawn due to the servers render distance being set to 6, but that's just a shot in the dark though. we're running a pretty hefty forge pack so it might be conflicting with one of the mods. The fact that there where only that many at one specific area is confusing though, although almost everyone else's base also has about 100 floating around.

We're running the 1.7.1 version, if you wanna know anything more then i'm happy to provide more info.

**Comment by Tschipcraft** [comment]
90,000 is a bit... _much_. Even 100 at one base should not be normal...
A few questions:
1. Are multiple markers at the exact same position?
2. Are light blocks at every marker?
  2.1 What are the tags of the markers where this is not the case (specifically, if they have the tag ts.dl.remove)?
3. Are all light blocks actually associated with something that should emit light? (Or in other words, are there light blocks somewhere where they shouldn't be)

The only cause I can think of right now is maybe a save desync like you mentioned caused by a performance optimization mod or similar, but I'm not sure.

**Comment by Ninty9** [comment]
I made a backup before killing all the marker entities but its gonna take a bit to grab from the server, but i do remember that there where multiple markers at the same position. I'll reply with more when i figure it out.

**Comment by Ninty9** [comment]
They all have the tags "global.ignore.pos", "global.ignore". "ts.dl.remove", "ts.dl.light:", "global.ignore.kill", "smithed.block". It's hard to be sure but there are definitely ones at the same x and z coordinates, and considering theres 100.000 of them theres probably ones at the same position.

Also there seems to be no light blocks lingering anywhere. so they did all get removed.

**Comment by Tschipcraft** [comment]
Thanks for the information!
It really looks like the issue is that light entity markers are spawned outside the world, then don't actually place a light block because the area is not loaded and just don't get despawned, even though they are scheduled to be (ts.dl.remove). I also do not test for already existing markers at a specific location if no light block is present for better performance, so that would also explain why there are multiple at the same position. I'll work on a solution for this.

**Comment by Tschipcraft** [comment]
Just a quick question, the <kbd>:</kbd> at the end of the `ts.dl.light` tag was a typo on your end, right?

**Comment by Ninty9** [comment]
Yes, that was a typo. Something else that may or may not be important is that i had chunkloaders in the area with a large amount of markers, though im not sure what would be making the light blocks appear in that case but it might be a factor

**Comment by Tschipcraft** [comment]
Interesting..
Well I'm still struggling to find the underlying cause of this issue, could you perhaps send over the old backup of the world including a list of the mods used for me to experiment on?

**Comment by ExWhyZedd0** [comment]
hey, cus you ask for his world. here my world, if you want to try check it. so if someone loaded my storage room, the tps will go very low.

here the link to download it (i didnt save it locally the file anymore)
https://binusianorg-my.sharepoint.com/personal/felix_marcellino_binus_ac_id/_layouts/15/guestaccess.aspx?share=ElanrE3VBQ5NpMEHQnUL0RUB-EX3PrhfpN48y4dio69Zew&e=Du9Mkz

**Comment by Tschipcraft** [comment]
Thanks, I'll take a look at it

**Comment by Tschipcraft** [comment]
In your case @ExWhyZedd0, the problem just comes down to the around 200 glow item frames used in the storage area, since they are treated as dynamic light sources. I did not find any difference in performance between v1.6 and v1.7, so this might have just been a coincidence with the time when you built the area. When testing the latest commit in the current dev branch - which also contains some improvements from v1.7.1 - performance got around 8 ms better but was still 12 ms worse than without Dynamic Lights on my system. I may be able to optimize the performance even more, especially for item frames, since they normally don't move, in future updates.
Since marker entity numbers stayed consistent and were nowhere near 90.000, this is not exactly the same problem as @Ninty9 experienced, but may have contributed to the bad performance of course as well. I'd still be happy to take a look at your world, @Ninty9 to see what's going on.

**Comment by Tschipcraft** [comment]
Yes, please go ahead

**Comment by Tschipcraft** [comment]
Any updates on this @Ninty9?

**Comment by oakenduck** [comment]
As of v1.8.5/1.21.4, this issue is still ongoing. I'll notice heavy lag and then do a `/kill` of marker entities, resulting in hundreds of thousands of markers killed. If I do periodic cleanups, then light blocks get left at random places all over my world. Any fixes?

**Comment by Tschipcraft** [comment]
Please do not use `/kill` for markers. They keep track of where the light blocks were placed, so when they're killed, light blocks are left in the world. (Amount of markers always corresponds to amount of active dynamic lights in the world)
I'm not sure where so many lights are coming from though, might be best to just limit them in the next update. Do you have a lot of items or item frames in your world?
